static/app/components/tours/tourContext.tsx

```diff
@@ -208,6 +214,13 @@ export function useTourReducer<T extends TourEnumType>(
     dispatch({type: 'SET_STEP', stepId});
   }, []);
 
+  useEffect(() => {
+    if (state.currentStepId) {
+      options?.onStepChange?.(state.currentStepId);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state.currentStepId]);
```

Comment:
Here too, this can introduce stale closures: if  `options.onStepChange` is a function that gets re-created during a render, we might run a “stale version” of that function here.

The best fix would be to not invoke this callback in an effect at all, but to move the call to the event handlers that trigger the state change. If this is not reasonable because we need to compute the result first in the reducer, [the latest ref pattern](https://www.epicreact.dev/the-latest-ref-pattern-in-react) would also be good where we store the callback in a ref that updates on every render so that we can invoke it in the effect without adding it to the dependency array.

Reply:
Good call here! This _does_ depend on state changes from the reducer, but I decided to extract the step computation to util functions so that we can invoke the callbacks from the event handler.

Reply:
Okay that is incredible, I didn't know that existed but it would have made my life so much easier in a couple instances